File: SpeedTestGameClient/core/src/main/java/si/um/feri/speedii/dao/http/HttpMobileTower.java
package si.um.feri.speedii.dao.http;


import io.github.cdimascio.dotenv.Dotenv;
import si.um.feri.speedii.classes.Location;
import si.um.feri.speedii.classes.Measurement;
import si.um.feri.speedii.classes.MobileTower;
import si.um.feri.speedii.classes.SessionManager;
import si.um.feri.speedii.classes.Type;
import si.um.feri.speedii.classes.User;
import si.um.feri.speedii.dao.MeasurementCRUD;
import okhttp3.*;
import si.um.feri.speedii.dao.MobileTowerCRUD;

import org.bson.types.ObjectId;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class HttpMobileTower implements MobileTowerCRUD {
    private final OkHttpClient client;
    private final SessionManager sessionManager;
    private final String ip;

    public HttpMobileTower(SessionManager sessionManager) {
        this.sessionManager = sessionManager;
        this.client = new OkHttpClient();

        Dotenv dotenv = Dotenv.configure().filename(".env").load();
        this.ip = dotenv.get("IP");
    }

    @Override
    public List<MobileTower> getByConfirmed(boolean status) {
        String url = ip + "/mobile/confirmed/" + status;
        Request request = new Request.Builder()
            .url(url)
            .get()
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Failed to execute request: " + response.code());
                return new ArrayList<>();
            }

            String responseBody = response.body() != null ? response.body().string() : null;
            if (responseBody != null) {
                JSONArray jsonArray = new JSONArray(responseBody);
                List<MobileTower> mobileTowers = new ArrayList<>();

                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject jsonObject = jsonArray.getJSONObject(i);
                    mobileTowers.add(parseMobileTower(jsonObject));
                }

                return mobileTowers;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return new ArrayList<>();
    }



    @Override
    public boolean toggleConfirm(MobileTower obj) {
        String url = ip + "/mobile/confirm/" + obj.getId();
        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .get()
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (response.isSuccessful()) {
                obj.setConfirmed(!obj.isConfirmed());
                return true;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return false;
    }

    @Override
    public MobileTower getById(ObjectId id) {
        String url = ip + "/mobile/" + id;
        Request request = new Request.Builder()
            .url(url)
            .get()
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (response.isSuccessful()) {
                String responseBody = response.body() != null ? response.body().string() : null;
                if (responseBody != null) {
                    JSONObject jsonObject = new JSONObject(responseBody);
                    return parseMobileTower(jsonObject);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return null;
    }

    // TESTED
    public List<MobileTower> getByLocator(User user) {
        String url = ip + "/mobile";
        Request request = new Request.Builder()
            .url(url)
            .get()
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Failed to execute request: " + response.code());
                return new ArrayList<>();
            }

            String responseBody = response.body() != null ? response.body().string() : null;
            if (responseBody != null) {
                JSONArray jsonArray = new JSONArray(responseBody);
                List<MobileTower> mobileTowers = new ArrayList<>();

                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject jsonObject = jsonArray.getJSONObject(i);
                    MobileTower mobileTower = parseMobileTower(jsonObject);

                    if (mobileTower.getLocator() != null && mobileTower.getLocator().getId().equals(user.getId())) {
                        mobileTowers.add(mobileTower);
                    }
                }

                return mobileTowers;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return new ArrayList<>();
    }

    // TESTED
    @Override
    public List<MobileTower> getAll() {
        String url = ip + "/mobile";
        Request request = new Request.Builder()
            .url(url)
            .get()
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Failed to execute request: " + response.code());
                return new ArrayList<>();
            }

            String responseBody = response.body() != null ? response.body().string() : null;
            if (responseBody != null) {
                JSONArray jsonArray = new JSONArray(responseBody);
                List<MobileTower> mobileTowers = new ArrayList<>();

                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject jsonObject = jsonArray.getJSONObject(i);
                    mobileTowers.add(parseMobileTower(jsonObject));
                }

                return mobileTowers;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return new ArrayList<>();
    }

    @Override
    public boolean insert(MobileTower obj) {
        String url = ip + "/mobile";
        RequestBody requestBody = createRequestBody(obj);
        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .post(requestBody)
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return false;
    }

    @Override
    public boolean insertMany(List<MobileTower> list) {
        JSONArray jsonArray = new JSONArray();
        for (MobileTower tower : list) {
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("location", new JSONObject()
                .put("type", "Point")
                .put("coordinates", tower.getLocation().coordinates));
            jsonObject.put("operator", tower.getProvider());
            jsonObject.put("type", tower.getType());
            jsonObject.put("confirmed", tower.isConfirmed());
            jsonObject.put("locator", tower.getLocator() != null ? tower.getLocator().toString() : null);
            jsonArray.put(jsonObject);
        }

        JSONObject jsonTowers = new JSONObject().put("towers", jsonArray);
        RequestBody requestBody = RequestBody.create(jsonTowers.toString(), MediaType.get("application/json"));
        Request request = new Request.Builder()
            .url(ip + "/mobileTowerRoutes/createMany")
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .post(requestBody)
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return false;
    }

    @Override
    public boolean update(MobileTower obj) {
        String url = ip + "/mobile/" + obj.getId();
        RequestBody requestBody = createRequestBody(obj);
        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .put(requestBody)
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return false;
    }

    @Override
    public boolean delete(MobileTower obj) {
        String url = ip + "/mobile/" + obj.getId();
        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .delete()
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return false;
    }

    private RequestBody createRequestBody(MobileTower obj) {
        JSONObject json = new JSONObject()
            .put("location", new JSONObject()
                .put("type", "Point")
                .put("coordinates", obj.getLocation().coordinates))
            .put("operator", obj.getProvider())
            .put("type", obj.getType())
            .put("confirmed", obj.isConfirmed());
        if (obj.getLocator() != null) {
            json.put("locator", obj.getLocator().getId());
        }

        return RequestBody.create(json.toString(), MediaType.get("application/json"));
    }

    private MobileTower parseMobileTower(JSONObject jsonObject) {
        JSONObject locationObject = jsonObject.getJSONObject("location");
        JSONArray coordinates = locationObject.getJSONArray("coordinates");
        Location location = new Location(List.of(coordinates.getDouble(0), coordinates.getDouble(1)));
        String operator = jsonObject.getString("operator");
        String type = jsonObject.getString("type");
        boolean confirmed = jsonObject.getBoolean("confirmed");
        User locator = jsonObject.isNull("locator") ? null : parseUser(jsonObject.getJSONObject("locator"));
        ObjectId id = jsonObject.has("_id") ? new ObjectId(jsonObject.getString("_id")) : new ObjectId();

        return new MobileTower(location, operator, type, confirmed, locator, id);
    }

    private User parseUser(JSONObject jsonObject) {
        String username = jsonObject.getString("username");
        String password = jsonObject.getString("password");
        String email = jsonObject.getString("email");
        boolean admin = jsonObject.getBoolean("admin");
        ObjectId id = new ObjectId(jsonObject.getString("_id"));

        return new User(username, password, email, admin, id);
    }
}

File: MobileApp/app/speedtest/src/main/java/sosteric/pora/speedtest/dao/http/HttpMobileTower.kt
package dao.http

import Location
import Measurment
import MobileTower
import SessionManager
import User
import android.graphics.Bitmap
import android.util.Base64
import android.util.Log
import dao.MobileTowerCRUD
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.bson.types.ObjectId
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File

class HttpMobileTower(val sessionManager: SessionManager) : MobileTowerCRUD{
    private val client = OkHttpClient()
    var ip = "http://20.160.43.196:3000"

    override fun getByConfirmed(status: Boolean): List<MobileTower> {
        val url = "${ip}/mobile/confirmed/${status}"
        val request = Request.Builder()
            .url(url)
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                // If not, handle the error
                println("Failed to execute request: ${response.code}")
                return emptyList()
            }
            val responseBody = response.body?.string()

            responseBody?.let {
                val jsonArray = JSONArray(it)
                val mobileTowers = mutableListOf<MobileTower>()

                for (i in 0 until jsonArray.length()) {
                    val jsonObject = jsonArray.getJSONObject(i)

                    mobileTowers.add(parseMobileTower(jsonObject))
                }

                return mobileTowers
            }

            return emptyList()
        }
    }

    override fun toggleConfirm(obj: MobileTower): Boolean {
        val url = "${ip}/mobile/confirm/${obj.id}"
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .get()
            .build()


        client.newCall(request).execute().use { response ->
            if(response.isSuccessful){
                obj.confirmed.not()
                return true
            }
            return false
        }

    }

    override fun getById(id: ObjectId): MobileTower? {
        val url = "${ip}/mobile/$id"
        val request = Request.Builder()
            .url(url)
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                responseBody?.let {
                    val jsonObject = JSONObject(it)
                    return parseMobileTower(jsonObject)
                }
            }
            return null
        }
    }


    fun getByLocator(user: User): List<MobileTower> {
        val url = "$ip/mobile"
        val request = Request.Builder()
            .url(url)
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                println("Failed to execute request: ${response.code}")
                return emptyList()
            }

            val responseBody = response.body?.string()

            responseBody?.let {
                val jsonArray = JSONArray(it)
                val mobileTowers = mutableListOf<MobileTower>()

                for (i in 0 until jsonArray.length()) {
                    val jsonObject = jsonArray.getJSONObject(i)
                    val mobileTower = parseMobileTower(jsonObject)

                    // Filtriraj towerje glede na locator
                    if (mobileTower.locator?.id == user.id) {
                        mobileTowers.add(mobileTower)
                    }
                }

                return mobileTowers
            }

            return emptyList()
        }
    }


    override fun getAll(): List<MobileTower> {
        val url = "$ip/mobile"
        val request = Request.Builder()
            .url(url)
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                println("Failed to execute request: ${response.code}")
                return emptyList()
            }

            val responseBody = response.body?.string()

            responseBody?.let {
                val jsonArray = JSONArray(it)
                val mobileTowers = mutableListOf<MobileTower>()

                for (i in 0 until jsonArray.length()) {
                    val jsonObject = jsonArray.getJSONObject(i)

                    mobileTowers.add(parseMobileTower(jsonObject))
                }

                return mobileTowers
            }

            return emptyList()
        }
    }
    override fun insert(obj: MobileTower): Boolean {
        val url = "$ip/mobile"
        val requestBody = createRequestBody(obj)
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .post(requestBody)
            .build()

        client.newCall(request).execute().use { response ->
            Log.d("AddTower", "Response: ${response.body?.string()}")
            return response.isSuccessful
        }
    }

    fun confirm(bitmap: Bitmap): Int {
        try {
            val bos = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, 100, bos)
            val bitmapData = bos.toByteArray()

            val requestBody = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("image", "tower_image.jpg", bitmapData.toRequestBody("image/jpeg".toMediaTypeOrNull()))
                .build()

            val request = Request.Builder()
                .url("$ip/mobile/addConfirm")
                .addHeader("authorization", "Bearer ${sessionManager.token}")
                .post(requestBody)
                .build()

            client.newCall(request).execute().use { response ->
                if (response.isSuccessful) {
                    val responseBody = response.body?.string()
                    responseBody?.let {
                        println("Mobile tower confirmed: $it")

                        val jsonResponse = JSONObject(it)
                        val data = jsonResponse.getJSONObject("data")
                        val confirmed = data.getBoolean("confirmed")

                        return if (confirmed) { 1 } else { 0 }
                    }
                } else {
                    Log.d("AddTower", "Failed to confirm mobile tower. Response: ${response.code}")
                    println("Failed to confirm mobile tower. Response: ${response.code}")
                    return -1
                }
            }
        } catch (e: Exception) {
            Log.d("AddTower", "Failed to confirm tower. Exception: $e")
            e.printStackTrace()
            return -1
        }
        return -1
    }




    override fun insertMany(list: List<MobileTower>): Boolean{
        val jsonArray = JSONArray()
        list.forEach { tower ->
            val jsonObject = JSONObject().apply {
                put("location", JSONObject().apply {
                    put("type", "Point")
                    val coordinatesArray = JSONArray()
                    tower.location.coordinates.forEach { coordinatesArray.put(it) }
                    put("coordinates", coordinatesArray)
                })
                put("operator", tower.provider)
                put("type", tower.type)
                put("confirmed", tower.confirmed)
                put("locator", tower.locator.toString())
            }
            jsonArray.put(jsonObject)
        }

        val jsonTowers = JSONObject().apply {
            put("towers", jsonArray)
        }

        val requestBody = jsonTowers.toString().toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url("${ip}/mobileTowerRoutes/createMany")
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .post(requestBody)
            .build()

        client.newCall(request).execute().use { response ->
            return response.isSuccessful
        }
    }

    override fun update(obj: MobileTower): Boolean {
        val url = "$ip/mobile/${obj.id}"
        val requestBody = createRequestBody(obj)
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .put(requestBody)
            .build()

        client.newCall(request).execute().use { response ->
            return response.isSuccessful
        }
    }

    override fun delete(obj: MobileTower): Boolean {
        val url = "$ip/mobile/${obj.id}"
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .delete()
            .build()

        client.newCall(request).execute().use { response ->
            return response.isSuccessful
        }
    }
    private fun createRequestBody(obj: MobileTower): okhttp3.RequestBody {
        val json = JSONObject().apply {
            put("location", JSONObject().apply {
                put("type", "Point")
                val coordinatesArray = JSONArray()
                obj.location.coordinates.forEach { coordinatesArray.put(it) }
                put("coordinates", coordinatesArray)
            })
            put("operator", obj.provider)
            put("type", obj.type)
            put("confirmed", obj.confirmed)
            if(obj.locator != null) {
                put("locator", obj.locator!!.id)
            }
        }
        return json.toString().toRequestBody("application/json".toMediaType())
    }

    fun parseMobileTower(jsonObject: JSONObject): MobileTower {
        val locationObject = jsonObject.getJSONObject("location")
        val coordinates = locationObject.getJSONArray("coordinates")
        val location = Location(coordinates = listOf(coordinates.getDouble(0), coordinates.getDouble(1)))
        val operator = jsonObject.getString("operator")
        val type = jsonObject.getString("type")
        val confirmed = jsonObject.getBoolean("confirmed")
        val locator = if (!jsonObject.isNull("locator")) {
            parseUser(jsonObject.getJSONObject("locator"))
        } else {
            null
        }
        val id = if (jsonObject.has("_id")) ObjectId(jsonObject.getString("_id")) else ObjectId()
        return MobileTower(location, operator, type, confirmed, locator, id)
    }
    fun parseUser(jsonObject: JSONObject) : User {
        val username = jsonObject.getString("username")
        val password = jsonObject.getString("password")
        val email = jsonObject.getString("email")
        val admin = jsonObject.getBoolean("admin")
        val id = ObjectId(jsonObject.getString("_id"))
        return User(username, password, email, admin, id)
    }


}